### PR TITLE
refactor!: give items identity

### PR DESCRIPTION
**Pull request by joveeater**
#### Summary

SUMMARY: Infrastructure "Support managed memory and safe references for items"

#### Purpose of change

Give items identity. This is the start of a big refactor. I want to pay off some technical debt, untangle the spaghetti and get all our ducks in a row. The real aim is ease of development through consistency and safety. I've spoken about this a bit before but eventually this will mean moving more things to actors allowing the rest of the engine to focus on specific clear tasks and not have to worry about unexpected invalidations. Actors themselves need to be safe and so the first step is safe references. I want to call this new system game objects or GOs. To begin with that will mean items but creatures and vehicles are the next targets. Maybe it could stretch to something like fields but not much further. The main aim at the moment is just to get nice consistent interfaces that we can build out behind later. 

GOs know their own location and allow you to do things like describe that location or remove them from it. This is kept up to date automatically as the object moves around the game world. GOs can't be constructed normally (private constructors), instead they use static `::spawn` methods. The main `::spawn` will give you a GO without a location and it's your duty to give it one or call `->destroy()` before the end of the turn. `::spawn_template` and `_temporary` give you a GO with a special location that will debugmsg if you try to put it in the game world. Temporaries are destroyed at the end of turn. Note that a GO is "loaded" if it's inside the reality bubble as such templates/temporaries are never loaded.

GOs cannot have two locations at once and they can't be destroyed while still having a location. Attempting to do either will debugmsg and then call `detach()` to remove them from their location first. They debugmsg to warn you to be careful of iterator invalidation. GO containers should provide methods that update both iterators and locations correctly. If iterators aren't a problem just call detach yourself first. Note that detaching something infers the same responsibility to put it somewhere or destroy it that creating something new does. Failing to do this will cause debugmsgs in debug builds and memory leaks in release. This level of strictness around iterators might not be appropriate for monsters/vehicles etc and we can just cut that debugmsg but I think it's warranted for items. 

Normal references and pointers to GOs are always valid until the end of the turn, even through movement/destruction/unloading. This means that when appropriate it's important to check their destroyed/loaded status, however with good actor separation that should only be when you're directly engaged in destroying stuff etc. Conversely they should never be considered valid across turns so they're suitable for the temporary variables/structures that get passed around as the engine does its work. If you want a reference that goes across turns you need one of two depending. A `cache_reference` is safe, it will deny access to unloaded or destroyed things and it's fast, but it can't be stored in save files and once something leaves the reality bubble it loses track of it permanently so it's best for caches. A `safe_reference` is slower but can be serialized and allows explicit const access to things that were destroyed/unloaded this turn and doesn't lose track of its target. They can also tell the difference between destroyed and unloaded targets, which cache references can't.

And that's all you really need to know to use GOs and for the most part they all work the same, but this is a cleanup effort. More important than what's added is what's being removed. The old `safe_reference` is first on the block and `item_location` is the same. Half of its features are going in items themselves and the other half in references. Note that new item locations work similarly to the old ones, except they're a bit more granular wrt things like wielded/worn and of course they're kept up to date by methods like add_item_or_charges or rem_i etc. Other GOs will have a similar system of locations as well but this is less of a change for them.  

#### Describe the solution

Most of the implementation is fairly simple if extensive. I said pointers were only used for turn to turn stuff but actually they're used internally for whatever's storing an object's location as well. This is a big change for items as there's an extra layer of indirection on most of their structures. Something like a vector of item is a compile error now with the private constructors, it needs to be item*. This means lots of syntax changes where . becomes -> etc, not to mention the many constructors becoming `::spawn` and since I'm going through all of it anyway I'm going to take the time to look at the structures used for all this, move everything across to vectors for now and then come back and do some optimization in important places. 

Safe references are not simple, mostly because I want something that could last in the json for compatibility down the line and that has good save scumming properties. Rather than go into detail on the implementation here check the comments in safe_reference.h but I still need to finish, test and debug it as with all the code. The idea is that save scumming can have some effects but they're limited. References to things that now never existed are perpetually unloaded. Otherwise it will usually only cause small amounts of memory to be leaked or safe references to destroyed things to tell you they're unloaded or vice versa and typically code won't care to draw a distinction between those two states anyway. In rare cases where objects that are merged together both already have an internal id it can cause one half to point to a perpetually unloaded object. 

Internal ids used by safe references aren't exposed and are generated lazily. I'm gonna take a look at their format but for now they're using the fact that they're always generated as the result of a save and just storing the timestamp of that save as the start of the id. Really this will probably be enough, so long as you're not messing with your system clock collisions shouldn't be a problem. In the actual json GOs just have an optional id field, once they have an id they don't lose it, but it's stored externally to the GO itself in memory. Most GOs don't have one at all and then safe references in json are just a numeric id. 

As you can see from the mess that is the current code there's still a lot to do. Next step is taking inventory of all the places an item can be in the game world, doing a location for each and restricting changes except through methods that update location. Then fix all the todos and errors and generally clean everything up, resolve conflicts, test the shit out of it and get this thing in a mergeable state. There'll probably be a lot of teething problems but at least things like activities should be able to track items more reliably. Then it's back for creatures and vehicles becoming GOs, then doing better actors, then character stats/abilities at which point we should have good interfaces that we're happy to expose to LUA and support going forward and I will finally have achieved the initial goal that led me down this path of not having to hardcode stuff against traits.  

Then the real work can begin and we can start improving what's behind those interfaces. It's all a long way off but I want to make a real push towards getting a solid consistent engine that's easy to work with and debug. This first step with items is by far the hardest but consequently the most pressing. None of the code changes are at all final, they all need checking for correctness. They just show the general direction and extent of what's needed. 

#### Describe alternatives you've considered

Myriad. More important than any specific plan is consistently applying that plan across the engine and there are still a million decisions to make. Next steps for me is I'm gonna create one of those fancy projects that lays out the wider plan. Then hack away at this whole thing until it's done. I'll be detouring for stuff like the renderer at some point and once the interfaces are nicely in place I do wanna come back to jumping and flight, but I'm gonna try to focus on these big architectural issues. I thought our biggest problems were `g->m` and npc ai. What a grand and intoxicating innocence. 

Anyway, this is the first of probably many greats walls of text, not to mention this all needs good documentation. Let me know what you think. Any specifics are still up in the air.

### CODE REVIEW

**Comment by joveeater** [comment]
Right, more explaining the new location types. I've started to update the docs although some stuff is still missing. The big snarl to this is that it means all of the structures representing physical locations within the world need their types changing slightly. The new location_vector doesn't quite meet the interface for a vector, because it uses different types for adding/removing than it does for examining. I was hoping we could get away with a few simple types like vector but crafting vs real inventories means that needs duplicating too.

The real workhorse of this scheme is the detached_ptr type. detached_ptr is used to represent items that aren't currently in the world. You need to use it with std::move, which prevents any item duplication and if you don't move it the item will get destroyed when the detached_ptr does. It's very similar to a unique_ptr except that it's only doing our soft destroy, not a real delete which will happen at the end of the turn. 

This means a bit of familiarity with std::move and rvalue references when writing code that moves items around. For the most part it just means you have to remember to use std::move when passing a detached_ptr somewhere and you get a compile error if not. When you're writing functions that take a detached_ptr (i.e. functions that might place it somewhere in the world, use a normal pointer/reference otherwise), you should use detached_ptr<item>&& in your signature. Then, the caller can check the detached_ptr as a boolean afterwards to see if the item was actually placed in the world and run a fallback and whatever they do it'll be destroyed correctly.

Note that a lot of item moving functions, for instance character::wield, still don't take a detached_ptr, because they deal with detaching the item themselves. You can easily get a detached_ptr from an item by calling detach(), but this removes it from the world and invalidates iterators etc. It also gives red text and fails if the item is already detached in order to maintain the only one detached_ptr at a time rule. All of the item removal functions now also return a detached_ptr. 

There's a bit of fake item marshaling. Fake items are split into two categories. There are temporary items for things that're used and then immediately discarded and they're identical to spawning something then not moving the detached_ptr you get anywhere. There's also actual fake items that use a special fake item location, they have to live on actors and they won't let you detach them. Speaking of, the memory management around actors and activities is changing a little. They're now more consistently using unique_ptr and doing less copying of themselves which means we don't have to worry about copying their fake items.

What I'd like to do at some point is get rid of the null item as well or at least get rid of all its legitimate uses and rename it to the error item. I'm not gonna do that yet, but I am gonna use the null item as a cushion for memory errors. Attempting to dereference any of the new pointer types will give the null item and red text rather than an actual nullptr so we can mitigate crashes.

**Comment by joveeater** [comment]
I am marking this as ready for review. Since this is so big we should probably forgo the normal process and wait until everyone has had a chance to look at it. There's still a bit more to clean up but unless there are any objections I'm hoping to merge next Friday (28th) so I'll have the weekend to handle any problems that come up. 

This is 7 months of work finally coming together but I'm hoping to press right on afterwards with codegen stuff. I'm gonna start with some of the huge constructors that are needed for the location stuff but hopefully I can make something reusable.

**Comment by scarf005** [comment]
@joveeater could you also provide some tutorial docs for beginners?
It'd be really helpful to have simple snippets that shows how to use GameObjects.
Some practical examples may be (it can be much simpler, below are just a general idea):

- sort foods by their nutrients, take first 3, and eat it (delete them)
- when eating food, find the location of the food, and if it's cold (e.g in fridge), give morale bonus
- find wielded gun, unload empty magazine, find a magazine from inventory, load it
- map through all items in inventory, and damage them
- example test that finds all items near 3 tiles radius and confirm they're fresh

**Comment by joveeater** [comment]
@scarf005 That's a great idea! I'll give a rough idea of what I'd do for the ones you've suggested here and probably tweak them a bit for the docs.

First though, take a look at visitable.h. It's got functions for mapping/filtering over items and it's supported by a whole bunch of things (tiles, characters, vehicles etc). It existed before this PR but it's had an update. A lot of the functions I'll talk about are part of this interface. Note that it also solves the problem that "map over items in inventory" prompts the question "would you like to map over liquids in containers and bullets in guns etc?"

It's good to understand the `detached_ptr&& -> detached_ptr` functions as well. Most of the functions that place an item in the world work like this now and I'll look at how to really use them in the examples. For now, they will return whatever they don't handle (i.e. take ownership of by placing or destroying). So if you try to put an item in an inventory and it's full, it will return the item back to you. If they do handle the item you get an empty detached_ptr back and they silently do nothing when passed an empty detached_ptr, more on that later.

On to the examples, for the first one, eating food by nutrients. Maybe start with the player's crafting inventory? Inventories are visitable, so we can use `items_with` to get a `vector<item*>` of any food. `std::sort` lets us do the calories bit and then we can use `avatar_action::eat` which takes an `item*` and acts as if you just pressed 'e' on it. All the deleting the item bit is handled for you that way, though if you want to do that yourself you can use `it->detach()` and go from there. If you just ignore the result of detach the item will be deleted.

The second one, morale bonus for cold. Items now support `tripoint x=item->position();` and you should be able to go from there. There's also `item->where();` which will give you an enum telling you if it's on the ground etc. If you're going to detach the item make sure you call these functions before you do, detached items don't have a position and will red text you if you try.

The third one, loading a gun. You can kinda just do it. `mag->detach()` and whatever the gun unload function is called will both remove the mag and give you the detached_ptr to it. Then you put it in the player's inventory etc, but that put in inventory function is `detached_ptr&& -> detached_ptr`. If you just ignore the return value the mag will get deleted when your inventory is full. 

This is where the idea that these functions silently ignore empty detached_ptrs becomes useful. If you try to put the mag in the player's inventory and then put whatever that returns on the floor, it acts as a fallback. If the inventory is full it goes on the floor. There's also the `attempt_detach( detached_ptr&& -> detached_ptr )` function, which lets you use "wherever it came from" as the final fallback, if the lambda you give it returns the item (i.e. all other options failed), it will just leave it where it is. 

For the fourth and fifth, visitable has the functions you need. Note most of the filter functions work like `attempt_detach` they take a `detached_ptr&& -> detached_ptr` function and will remove it if you return empty. 

Also, small update. We want to wait until 0.3 stable is released before merging this so I'm putting it back to a draft for now. There's still a few things left to do anyway, like adding examples to the docs. Thanks again scarf!

File: src/visitable.cpp [hunk 1]
```diff
@@ -375,6 +389,31 @@ VisitResponse visitable<T>::visit_items( const std::function<VisitResponse( item
...
+static VisitResponse visit_internal( std::function < VisitResponse( detached_ptr<item> &&e ) >
+                                     filter, location_vector<item> &items )
+{
+    VisitResponse last = VisitResponse::NEXT;
+    items.remove_with( [&last, &filter]( detached_ptr<item> &&e ) {
+        if( last == VisitResponse::ABORT ) {
+            return std::move( e );
+        }
+        last = filter( std::move( e ) );
+        if( last == VisitResponse::NEXT && e ) {
```
Comment (olanti-p) [inline src/visitable.cpp:403]: Use after move

File: tests/basecamp_test.cpp [hunk 2]
```diff
@@ -23,24 +23,25 @@ static const itype_id meat_cooked_id( "meat_cooked" );
...
-            can = item( sealed_can_id, calendar::turn_zero );
+            can = item::spawn( sealed_can_id, calendar::turn_zero );
             break;
         case canned_status::canned_unsealed:
-            can = item( unsealed_can_id, calendar::turn_zero );
+            can = item::spawn( unsealed_can_id, calendar::turn_zero );
             break;
     }
-    REQUIRE( can.contents.insert_item( food ).success() );
+    REQUIRE( can->contents.insert_item( std::move( food ) ).success() );
```
Comment (olanti-p) [inline tests/basecamp_test.cpp:44]: Tidy gives use-after-move here, probably something to do with how it occurs inside a Catch2 macro
https://github.com/cataclysmbnteam/Cataclysm-BN/actions/runs/5621692607/job/15232955602?pr=2250#step:6:2320

**Review by olanti-p — COMMENTED** [review]
Skimmed the PR and picked off some low-hanging fruit (CI warnings, typos, build speedup from removing heavier includes). Most of the suggestions are massive, so I've opened a PR: https://github.com/joveeater/Cataclysm-BN/pull/4

A more in-depth look gonna take some time, there's a _lot_ of code.

**Comment by joveeater** [comment]
@olanti-p Thank you for this. This is amazing! 

By the looks of it, some of the use after moves are false positives. There are a few functions that want to use their return value for something else and so they use their detached_ptr&& argument as an in/out parameter. I'll add the nolint stuff. I'm tempted to shuffle them about and make their current return values into out parameters instead, but there's only a few functions like this in the first place.

That said, I'll have to look at the catch one closer, it does do some weird stuff, and `merge_item` should probably just be a `detached_ptr&&->detached_ptr` function rather than returning a bool.

File: src/locations.h [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+class submap;
+class vehicle;
+class monster;
+class npc;
+struct tripoint;
+template<typename T>
+class detached_ptr;
+
+template<class T>
+class location
```
Comment (olanti-p) [inline src/locations.h:21]: Could use a comment or link to the docs.

File: src/location_vector.h [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+
+//TODO!: This should probably just be a static swap now and leave the std version alone.
+namespace std
+{
+template<typename T>
+void swap( location_vector<T> &lhs, location_vector<T> &rhs );
+}
+
+template<typename T>
+class location_vector
```
Comment (olanti-p) [inline src/location_vector.h:29]: Could use a comment or link to the docs.

**Review by olanti-p — COMMENTED** [review]
Opened another PR with suggestions:
https://github.com/joveeater/Cataclysm-BN/pull/5

Found a few bugs during a playtest:
1. After debug killing all monsters, sometimes debug messsages show up complaining about detached ptr and items on tiles (monster out of bounds of reality bubble? will have to check further)
2. Equipping item via "Armor" menu (`Num+` -> `E`) gives a debugmsg about iterators
3. Taking off a worn item via "`T`ake off" menu fails with "You don't have that item." The item promptly disappears, with no debugmsg.

File: src/avatar_action.cpp [hunk 5]
```diff
@@ -835,35 +838,35 @@ bool avatar_action::eat_here( avatar &you )
...
-            add_msg( _( "You leave the empty %s." ), it->tname() );
-        } else {
-            loc.remove_item();
+    } else {
+
+        loc->attempt_detach( [&you]( detached_ptr<item> &&it ) {
+            return you.consume_item( std::move( it ) );
+        } );
+        if( loc->is_food_container() || !you.can_consume_as_is( *loc ) ) {
+            add_msg( _( "You leave the empty %s." ), loc->tname() );
```
Comment (olanti-p) [inline src/avatar_action.cpp:861]: I wanted to say that after food has been consumed, the container item ceases to be a `is_food_container()`, so no message would be produced. Then I went into the game to actually test it, and it turned out it's outright impossible to consume an item in a container that's lying on the ground near the character: the log says that you've consumed the item and left behind the wrapper, while in fact nothing happens at all.

Example item: cheeseburger debug-spawned with its default container.

File: src/avatar_action.cpp [hunk 6]
```diff
@@ -997,12 +988,9 @@ void avatar_action::use_item( avatar &you, item_location &loc )
...
         } else {
-            const int obtain_cost = loc.obtain_cost( you );
-            loc = loc.obtain( you );
-            if( !loc ) {
-                debugmsg( "Failed to obtain target item" );
-                return;
-            }
+            const int obtain_cost = loc->obtain_cost( you );
+            //TODO!: check
+            loc->obtain( you );
```
Comment (olanti-p) [inline src/avatar_action.cpp:993]: There's some weird shenanigans with activatable count-by-charges items, could be related to this snippet.

Case 1:
1. Have 10 aspirin on ground and 10 aspirin in inventory
2. Stand nearby, `a`ctivate -> aspirin on ground
3. The aspirin in inventory is consumed instead

Case 2:
1. Have 10 aspirin on ground
2. Stand nearby, `a`ctivate -> aspirin on ground
3. Notice you now have `aspirin (-1)` in your inventory. The aspirin on ground is unchanged.

**Review by olanti-p — COMMENTED** [review]
Well, there go half the files that start with "a".
Also more playtesting:
- `U`nloading a yogurt from its plastic cup gives multiple debugmsgs and doesn't work.
- `a`ctivating a worn holster gives a debugmsg about empty location vector (doesn't seem to be affected by whether you have any pistols or not).

File: src/active_item_cache.cpp [hunk 7]
```diff
@@ -74,16 +70,21 @@ std::vector<item_reference> active_item_cache::get()
...
+
+        //TODO!: This is a bit of a hack, at the very least check the maths
+        //Skip the entries that were processed last turn.
+        it += num_to_process * ( to_turn<int>( calendar::turn ) % kv.first );
+
         for( ; it != kv.second.end() && num_to_process >= 0; ) {
-            if( it->item_ref ) {
-                items_to_process.push_back( *it );
+            if( *it ) {
+                items_to_process.push_back( & **it );
```
[the remark this answers is not in the document]
Comment (olanti-p) [reply src/active_item_cache.cpp:87]: Another issue with vanishing food:
1. Have 2 different characters in different game worlds
2. Load character A, debug spawn a raw potato (it has a shelf life of 1 year), save and quit
3. Without closing the game, load character B, then immediately save and quit
4. Without closing the game, load character A and take 1 step. The potato vanishes.

File: src/location_ptr.h [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+#include <memory>
+
+template<typename T>
+class location;
+
+template<typename T>
+class detached_ptr;
+
+template<typename T, bool error_if_null = true>
+class location_ptr
```
[the remark this answers is not in the document]
Comment (joveeater) [reply src/location_ptr.h:14]: Yeah I think I'm gonna link all these to a locations doc and keep the game objects doc focused on detached_ptrs which are the only ones you need for manipulating stuff. Locations are needed when you're adding a new place that something can be. So a location_ptr always stores a location and also optionally the item that's in that location. They make sure that an item's idea of its location is accurate. location_vector and location_inventory do the same thing but with a vector and inventory of items instead. It's a bit awkward having to do structures like that and having to initialize them with their location is the reason we can't use default constructors but once it's constructed it's all automatic. We can be statically sure that items know where they are.

File: src/activity_actor.cpp [hunk 9]
```diff
@@ -1191,42 +1204,36 @@ void throw_activity_actor::do_turn( player_activity &act, Character &who )
...
-    item thrown = *target.get_item();
-    if( target->count_by_charges() && target->charges > 1 ) {
-        target->mod_charges( -1 );
-        thrown.charges = 1;
-    } else {
-        target.remove_item();
-    }
-    ranged::throw_item( who, trajectory.back(), thrown, blind_throw_pos );
+    //TODO!:check
+    detached_ptr<item> det = target->count_by_charges() ? target->split( 1 ) : target->detach();
```
[the remark this answers is not in the document]
Comment (joveeater) [reply src/activity_actor.cpp:1212]: It's probably just me getting confused.

**Comment by olanti-p** [comment]
Found a CTD:
1. Spawn as a new character
2. Find zombies
3. Debug spawn "Cooking on a Budget" (or any other book, really)
4. `R`ead -> that book -> until you gain a level
5. Be prompted about zombies, choose to interrupt reading
6. `Esc` -> `S`ave and quit

That's the last for today. Bugs aside, knowing where the item comes from given only its pointer, and being able to grab it from there, is just _WOW_.

**Comment by joveeater** [comment]
I've picked off some of the easier ones. I'll have to come back for the rest tomorrow. Equipping via the armor menu no longer gives a debugmsg. Taking off no longer deletes the item. Two bugs affecting the asprin. One is that it wouldn't treat -1 as all and instead tried to actually split -1 charges and the other is that when using an item like that if the stack merged as you picked it up the item would be consumed from the old deleted stack rather than the newly merged one. You can consume things in containers again. Liquids now unload properly.

So there's still book crash, holsters, food weirdness and debug killing to look at. Thanks so much for these reports!

**Comment by olanti-p** [comment]
Can confirm fixes work.

Another issue with reading:
1. Find a zombie
2. Debug spawn book
3. Start reading book
4. Be interrupted by zombie
5. Save and quit
6. Load
7. Drop the book
8. Save and quit
9. In `.sav` file, there's now an entry in `safe_references` that isn't used anywhere.

Another issue with liquid unload:
1. Debug spawn yogurt
2. Unload -> Pour on the ground
3. The yogurt gets poured on the ground, but then you're prompted to deal with the yogurt again

Siphoning from cars doesn't work.

Trying to interact with tamed animals results in CTD.

Firing assault barbs mutation or finger laser bionic gives a debugmsg.

**Comment by joveeater** [comment]
2 more bugs. Rotated submaps aren't rotating vehicles properly and sometimes cache references go missing from the map. Sadly I don't have a way to reproduce that second one but I'll keep trying.

**Comment by joveeater** [comment]
Turns out both those bugs were the same cause, a bunch of stuff wasn't being rotated properly.

**Review by scarf005 — DISMISSED** [review]
image

items inside container are not getting saved. to reproduce, spawn an ice cream, save and exit then reload immediately. only plastic tub (container) is loaded.

this may be related with the fact debug-spawned items are not marked as `is_loaded`.

**Comment by joveeater** [comment]
> items inside container are not getting saved. to reproduce, spawn an ice cream, save and exit then reload immediately. only plastic tub (container) is loaded.
I couldn't reproduce this. Which version was it on? There was a bug where foods were decaying at the wrong time and it was often triggered by reloading. 

> this may be related with the fact debug-spawned items are not marked as is_loaded.
Once they're spawned in they are loaded. It's the version that you preview when you're in the debug menu that isn't loaded.

**Review by scarf005 — CHANGES_REQUESTED** [review]
https://github.com/cataclysmbnteam/Cataclysm-BN/assets/54838975/6ae0a2fa-5ef9-4a62-ac4b-d43b21aaffdb

found out an way to reliably reproduce the [issue](https://github.com/cataclysmbnteam/Cataclysm-BN/pull/2250#pullrequestreview-1658684330) on commit 283f7b4c8c58a1033324323dfbb890209814cb84.

## steps to reproduce
1. in debug menu, spawn an ice cream.
2. put it on floor.
3. save and exit.
4. re-load the world.
5. ice cream is gone, only its container (plastic tub) is left.

<details><summary>On save items are correctly serialized.</summary>

image

```c++
#include "catch/catch.hpp"

#include <sstream>
#include "json.h"
#include "submap.h"
#include "game_constants.h"
#include "int_id.h"
#include "point.h"
#include "type_id.h"

namespace
{

constexpr auto json = R"([
      0,
      0,
      [
        {
          "typeid": "plastic_bucket",
          "bday": 633682,
          "owner": "your_followers",
          "last_rot_check": 0,
          "drop_token": {
            "turn": 633682,
            "drop_number": 2,
            "parent_number": 2
          },
          "contents": {
            "items": [
              {
                "typeid": "icecream",
                "charges": 34,
                "bday": 633682,
                "active": true,
                "last_rot_check": 633682,
                "item_tags": ["COLD"]
              }
            ]
          }
        }
      ]
    ])";
} // namespace

TEST_CASE( "submap deserialize item with contents", "[submap][item][item_contents][serde]" )
{
    std::istringstream iss( json );
    JsonIn jsin( iss );
    submap sm( tripoint_zero );
    sm.load( jsin, "items", 28, tripoint_zero );

    const auto &items = sm.get_items( point_zero );
    REQUIRE( items.size() == 1 );

    const auto& bucket = items.front();
    CHECK( bucket->typeId() == itype_id( "plastic_bucket" ) );
    CHECK( !bucket->contents.empty() );

    const auto &contents = bucket->contents.all_items_top();
    CHECK( contents.size() == 1 );

    const auto &icecream = contents.front();
    REQUIRE( icecream->typeId() == itype_id( "icecream" ) );
}
```

submaps will also correctly load items. not sure what's causing this.

</details>

**Review by scarf005 — CHANGES_REQUESTED** [review]
## Error on disassembling

based on latest commit 283f7b4c8c58a1033324323dfbb890209814cb84.

https://github.com/cataclysmbnteam/Cataclysm-BN/assets/54838975/c7de3381-1ded-4e4f-ad51-3d20cfbbc81d

disassembling`(B)` a plastic tub with pocket knife will result in following error:
1. `Attempting to erase something from a vector with more than 1 active iterator`
2. `Attempting to push_back to a vector with active iterators`

doing it again will result in following error:
1. `Tried to find where of an item without a location`
2. `position called on [plastic tubs] without a position`
3. `Tried to detach [plastic tubs] not in a location.`
4. `Attempting to push_back to a vector with active iterators`

and despite the message 'you salvaged 27 plastic chunks`, it is nowhere to be seen.

[debug.log](https://github.com/cataclysmbnteam/Cataclysm-BN/files/12840003/debug.log)

P.S: could you enable issues on https://github.com/joveeater/Cataclysm-BN/ to organize issues regarding item identity?

**Review by scarf005 — DISMISSED** [review]

**Comment by scarf005** [comment]
@olanti-p @joveeater what should we do with clang-tidy errors here and there?

File: doc/GAME_OBJECTS.md
[the remark this answers is not in the document]
Comment (joveeater) [reply doc/GAME_OBJECTS.md:1]: Oh yeah this is a new doc added for the new features so it won't have been caught by the migration.

**Comment by joveeater** [comment]
I will have a go at the clang tidy errors later on today.

**Review by olanti-p — DISMISSED** [review]

**Review by scarf005 — DISMISSED** [review]

**Review by olanti-p — APPROVED** [review]

File: tests/vehicle_split_test.cpp [hunk 10]
```diff
@@ -81,5 +81,6 @@ TEST_CASE( "vehicle_split_section" )
         if( vehs.size() == 1 ) {
             CHECK( vehs[ 0 ].v->part_count() == 38 );
         }
+        break;
```
Comment (olanti-p) [inline tests/vehicle_split_test.cpp:84]: Related to the bug causing #3552 ?

**Pull request merged**